[chaotic-good] Client & server transport

ctiller opened the pull request:
Adapts work from https://github.com/grpc/grpc/pull/34728 and previous changes from @nanahpang, implements new v3 filter/transport interface, and brings up the core of the chaotic good transport.

The next change will bring a more complete test suite (for this and inproc).

Vignesh2208 commented on src/core/lib/transport/transport.cc:
```diff
@@ -327,11 +325,19 @@ void ForwardCall(CallHandler call_handler, CallInitiator call_initiator,
           call_handler.SpawnGuarded(
               "recv_trailing_metadata",
               [md = std::move(md), call_handler]() mutable {
+                gpr_log(GPR_INFO, "Pushing trailing metadata: %s",
```
Maybe this log should be removed ?

ctiller replied:
fixed

Vignesh2208 commented on src/core/ext/transport/chaotic_good/server_transport.cc:
```diff
@@ -0,0 +1,334 @@
+                               frame_header, std::move(buffers));
+                         });
+                   }),
+              Case(FrameType::kCancel,
```
When does the Cancel frame actually get sent ? I don't see a corresponding send in the ClientTransport. Is this something we plan to add later ?

Vignesh2208 reviewed (commented):
Did my first pass on the source files excluding the tests. Looks mostly good to me. Thanks for working on this craig :) Left a few minor clarifications

ctiller commented on src/core/ext/transport/chaotic_good/frame.cc:
```diff
@@ -50,53 +54,63 @@ const NoDestruct<Slice> kZeroSlice{[] {
+    header_.message_length = msg.length;
+    header_.message_padding = msg.padding;
+    output_.data = msg.message->payload()->Copy();
+    output_.data.Append(Slice::FromStaticBuffer(kZeros, msg.padding));
```
done

ctiller commented:
@Vignesh2208 thanks for the review! Addressed comments, and fixed the failing tests (a bug slipped into promise based endpoint that I've now addressed)

nanahpang commented on src/core/ext/transport/chaotic_good/chaotic_good_transport.h:
```diff
@@ -0,0 +1,113 @@
+    return TrySeq(
+        control_endpoint_->ReadSlice(FrameHeader::frame_header_size_),
+        [this](Slice read_buffer) {
+          auto frame_header =
```
Would the stack variable`frame_header` dtor before the `If` promise complete?

ctiller replied:
no, `If` guarantees to call promise factories before returning

nanahpang commented on src/core/ext/transport/chaotic_good/chaotic_good_transport.h:
```diff
@@ -0,0 +1,113 @@
+              frame_header.ok(),
+              [this, &frame_header] {
+                const uint32_t message_padding = std::exchange(
+                    last_message_padding_, frame_header->message_padding);
```
Is the value of `last_message_padding_` used somewhere else?

ctiller replied:
just here... std::exchange does a read/write operation on it

nanahpang reviewed (approved):
Thanks Craig for the works! I focus on the chaotic_good part mostly, but other parts looks good too. Left some minor comments.

ctiller commented on src/core/ext/transport/chaotic_good/frame_header.cc:
```diff
@@ -46,7 +46,6 @@ void FrameHeader::Serialize(uint8_t* data) const {
   WriteLittleEndianUint32(
       static_cast<uint32_t>(type) | (flags.ToInt<uint32_t>() << 8), data);
   if (flags.is_set(0)) GPR_ASSERT(header_length > 0);
-  if (flags.is_set(1)) GPR_ASSERT(trailer_length > 0);
```
it's possible to have zero lengths for these, so no

ctiller commented on src/core/ext/transport/chaotic_good/frame_header.cc:
```diff
@@ -70,11 +69,15 @@ absl::StatusOr<FrameHeader> FrameHeader::Parse(const uint8_t* data) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("Invalid message length: ", header.message_length));
+  }
+  if (header.flags.is_set(1) && header.message_padding <= 0) {
```
fixed

ctiller commented on src/core/ext/transport/chaotic_good/frame_header.cc:
```diff
@@ -70,11 +69,15 @@ absl::StatusOr<FrameHeader> FrameHeader::Parse(const uint8_t* data) {
-        absl::StrCat("Invalid trailer length", header.trailer_length));
+        absl::StrCat("Invalid message padding: ", header.message_padding));
   }
+  header.trailer_length = ReadLittleEndianUint32(data + 20);
```
no, 0 length is legal here (and negative is impossible)

Vignesh2208 commented on src/core/ext/transport/chaotic_good/chaotic_good_transport.h:
```diff
@@ -0,0 +1,111 @@
+                return Map(
+                    TryJoin<absl::StatusOr>(
+                        control_endpoint_->Read(frame_header->GetFrameLength()),
+                        TrySeq(data_endpoint_->Read(message_padding),
```
Can we optimize this slightly as follows ?
```
TrySeq(data_endpoint_->Read(message_padding + message_length), 
             [this, message_padding] (SliceBuffer buffer) {
                // Remove first message_padding bytes from buffer
             }
```

ctiller replied:
Let's measure first - I'm still expecting some problems getting the alignment story all sorted properly once we start testing the whole stack, and I think this form would be easier to iterate on.

Vignesh2208 commented on src/core/ext/transport/chaotic_good/server_transport.cc:
```diff
@@ -0,0 +1,332 @@
+          "server-write", [this, stream_id = frame_header.stream_id,
+                           call_initiator = *call_initiator]() {
+            return CallOutboundLoop(stream_id, call_initiator);
+          });
```
Should the call_initiator be stored into stream_map_ here ?

ctiller replied:
Yeah... it'll get fixed in the next PR.

Vignesh2208 reviewed (commented):
Did my second pass and left a few questions

Vignesh2208 reviewed (approved):
LGTM for the rest of it (apart from the 2 comments I had left before)
